File: USACO/2016-2017/Jan 2017/Plat/promote.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#define MAXN 100005

using namespace std;

struct Node {
	int mc, prof, pid, id;
};

bool operator <(Node a, Node b) {
	return a.prof > b.prof;
}


vector<Node> adj[MAXN], nodes;
int res[MAXN];
int N;
int pid = 1;
int BIT[MAXN];


void po(Node& n) {
	nodes[n.id].pid = pid;
	for(Node to: adj[n.id]) {
		pid++;
		po(to);
	}
	
	nodes[n.id].mc = pid;
}

int query(int a) {
	int sum = 0;
	while(a > 0) {
		sum += BIT[a];
		a -= a & -a;
	}
	
	return sum;
}

void update(int a) {
	while(a <= N) {
		BIT[a]++;
		a += a & -a;
	}
}

int main() {
	cin >> N;
	
	for(int i = 0; i < N; i++) {
		nodes.push_back(Node());
		nodes[i].id = i;
		cin >> nodes[i].prof;
	}

	for(int i = 1; i < N; i++) {
		int a;
		cin >> a;
		a--;
		adj[a].push_back(nodes[i]);
	}
	

	po(nodes[0]);
	sort(nodes.begin(), nodes.end());

	for(int i = 0; i < N; i++) {
		res[nodes[i].id] = query(nodes[i].mc) - query(nodes[i].pid);
		update(nodes[i].pid);
	}
	
	for(int i = 0; i < N; i++) cout << res[i] << endl;
}

File: USACO/2016-2017/Jan 2017/Plat/promote.java
// Promotion Counting

import java.util.*;
import java.io.*;

public class promote {
	static int[] BIT;
	static int pid = 1;
	
	public static void main(String[] args) throws IOException {
		String file = "";
		// BufferedReader br = new BufferedReader(new FileReader(new File(file + ".in")));
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		int N = Integer.parseInt(st.nextToken());
		BIT = new int[N+1];
		
		Node[] nodes = new Node[N];
		
		for(int i = 0; i < N; i++) {
			st = new StringTokenizer(br.readLine());
			nodes[i] = new Node(i);
			nodes[i].prof = Integer.parseInt(st.nextToken());
		}
		
		for(int i = 1; i < N; i++) {
			st = new StringTokenizer(br.readLine());
			nodes[Integer.parseInt(st.nextToken()) - 1].children.add(nodes[i]);
		}
		br.close();
		
		preOrder(nodes[0]);
		Arrays.sort(nodes, (a, b) -> b.prof - a.prof);
		
		int[] res = new int[N];
		
		for(Node n: nodes) {
			res[n.id] = find(n.mc) - find(n.pid);
			add(n.pid);
		}
		
		for(int i: res) System.out.println(i);
	}
	
	public static class Node {
		int mc, prof, pid, id;
		ArrayList<Node> children = new ArrayList<>();
		
		Node (int id) {
			this.id = id;
		}
	}
	
	public static void preOrder(Node n) {
		n.pid = pid;
		
		for(Node to: n.children) {
			pid++;
			preOrder(to);
		}
		
		n.mc = pid;
	}
	
	public static int find (int i) {
		int res = 0;
		while(i > 0) {
			res += BIT[i];
			i -= i & -i;
		}
		
		return res;
	}
	
	public static void add (int i) {
		while(i < BIT.length) {
			BIT[i]++;
			i += i & -i;
		}
	}
}
